driver/src/main/scala/org/mongodb/scala/internal/ExecutionContextObservable.scala

```diff
@@ -16,32 +16,59 @@
 
 package org.mongodb.scala.internal
 
-import scala.concurrent.ExecutionContext
+import java.util.concurrent.atomic.AtomicInteger
 
 import org.mongodb.scala.{Observable, Observer, Subscription}
 
-private[scala] case class ExecutionContextObservable[T](context: ExecutionContext, original: Observable[T]) extends Observable[T] {
+import scala.concurrent.ExecutionContext
+
+private[scala] case class ExecutionContextObservable[T](observable: Observable[T], context: ExecutionContext) extends Observable[T] {
 
+  // scalastyle:off method.length
   override def subscribe(observer: Observer[_ >: T]): Unit = {
-    original.subscribe(SubscriptionCheckingObserver(
+    observable.subscribe(SubscriptionCheckingObserver(
       new Observer[T] {
-        override def onError(throwable: Throwable): Unit = context.execute(new Runnable {
-          override def run() = observer.onError(throwable)
-        })
+        private val referenceCount = new AtomicInteger(1)
+        @volatile
+        private var error: Option[Throwable] = None
+        @volatile
+        private var onCompleteCalled = false
+
+        override def onSubscribe(subscription: Subscription): Unit = withContext(() => { observer.onSubscribe(subscription) })
 
-        override def onSubscribe(subscription: Subscription): Unit = context.execute(new Runnable {
-          override def run() = observer.onSubscribe(subscription)
-        })
+        override def onNext(tResult: T): Unit = {
+          referenceCount.incrementAndGet()
+          withContext(() => {
+            observer.onNext(tResult)
```

Comment:
Observer.onNext could throw? If so checkTerminated should go in a finally block

Reply:
contractually its not allowed to throw